### PR TITLE
Improved dev workflow

### PR DESCRIPTION
**Pull request by rauanmayemir**
Third time's a charm. 😆

- [x] Migrate to `esy`.
- [x] Upgrade `ocaml` to `4.06`.
- [x] Set up CI
- [x] Add a build status badge
- [x] More restructuring of the codebase. [Here's a good thread](https://discuss.ocaml.org/t/jbuilder-templates/958) with a project setup that almost floats my boat.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+## Contributing [TBD]
+
+Run `esy jbuilder runtest` to run the test suites.
+
+> If you're using `yarn`, running `yarn test` will give you nice colorized output.
+
+## Acknowledgements
+
+This project is directly based on the work of [@cristianoc](https://github.com/cristianoc) (`React Mini`) and [@whitequark](https://github.com/whitequark) (`opam-cross-*`).
```
Comment (wokalski) [inline README.md]: whitequark's work is a bit unrelated now. I think he'd better be credited somewhere where the cross compilation tooling belongs.

**Comment by wokalski** [comment]
@rauanmayemir `src/core/lib/` is unnecessary. `src/core/` is good enough IMO

**Comment by rauanmayemir** [comment]
But where are we going to put tests?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      - run: sudo npm -g install esy@preview
+      - restore_cache:
+          <<: *esy_cache_key
+      - run: esy install
+      - save_cache:
+          <<: *esy_cache_key
+          paths:
+            - ~/.esy
+      - run: esy build
+      - run: yarn test
```
Comment (wokalski) [inline .circleci/config.yml]: let's not make this detour through yarn. We should just run `esy jbuilder runtest --no-buffer` or have a `./test` script that does just that.

**Comment by wokalski** [comment]
how about `core/tests`?

**Comment by rauanmayemir** [comment]
@wokalski Ready for review. 😎

**Comment by rauanmayemir** [comment]
Yeah, we can add it when we start working on it. The important part was finding the best dir structure, `renderer_mac` will be exactly same as `renderer_term`.

**Pull request merged**